ui/raidboss/emulator/ui/Tooltip.ts

```diff
@@ -1,44 +1,103 @@
+const hideEvents = [
+  'mouseleave',
+  'blur',
+] as const;
+
+const validDirections = [
+  'top',
+  'right',
+  'bottom',
+  'left',
+] as const;
+type ValidDirection = typeof validDirections[number];
+
+abstract class HTMLTemplateElement extends HTMLElement {
+  abstract get content(): DocumentFragment;
+}
+
+const showEvents = [
+  'mouseenter',
+  'focus',
+] as const;
+
+type TemplatesType = {[Property in ValidDirection]: HTMLTemplateElement};
+
+const toPx = (px: number): string => `${px}px`;
+
 export default class Tooltip {
-  constructor(target, direction, text, autoShow = true, autoHide = true) {
-    if (!Tooltip.validDirections.includes(direction))
-      throw new Error('Invalid direction for tooltip: ' + direction);
+  private offset = {
+    x: 0,
+    y: 0,
+  };
+  private target: HTMLElement;
+  private direction: ValidDirection;
+  private tooltip: HTMLElement;
+  private inner: HTMLElement;
+  private arrow: HTMLElement;
 
-    if (typeof target === 'string')
-      target = document.querySelector(target);
+  private static templates: TemplatesType;
+
+  constructor(
+      targetRef: string | HTMLElement,
+      direction: ValidDirection,
+      text: string,
+      autoShow = true,
+      autoHide = true) {
+    Tooltip.initializeTemplates();
+
+    let target: HTMLElement | null;
+
+    if (typeof targetRef === 'string')
+      target = document.querySelector<HTMLElement>(targetRef);
+    else
+      target = targetRef;
+
+    if (!target) {
+      const msg = 'Invalid selector or element passed to Tooltip';
+      console.error(msg);
+      throw new Error(msg);
+    }
 
-    this.offset = {
-      x: 0,
-      y: 0,
-    };
     this.target = target;
     this.direction = direction;
-    // Technically, templates should use querySelector, but this gets called so often
-    // on trigger-heavy encounters that we should hack its performance a bit
-    this.tooltip = document.getElementById(`${direction}TooltipTemplate`)
-      .content.firstElementChild.cloneNode(true);
```

Comment:
Alternatively, could this just check if `firstElementChild` is an `HTMLTemplateElement` and `cloneNode` returns an `HTMLElement`, or is it worth the code to pre-cache them as you added here?

Reply:
The 'performance hack' here is using `getElementById` to lower the scope that `querySelector` would need to check against. Technically this could instead be:

```typescript
  static cloneTemplate(direction: ValidDirection): HTMLElement {
    const template = Tooltip.templates[direction];
    return template.content.querySelector('.tooltip') as HTMLElement;
  }
```

They're functionally the same.

For reference, `firstElementChild` is actually an `HTMLDivElement`.

Reply:
Yeah yeah, agreed than caching the `getElementById` lookup.  My original comment was mostly saying "you added a bunch of code to find and cache tooltip elements by name" and wondering if that was necessary vs the original code.